torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -159,17 +161,65 @@ void ProcessGroupAgent::join() {
   listenerThread_.join();
 }
 
+bool ProcessGroupAgent::checkNoPendingMessage() {
+  const auto worldSize = pg_->getSize();
+  std::vector<int64_t> cnts(2 * worldSize);
+  {
+    // copy recv message cnt to 0 - worldSize elements
+    std::lock_guard<std::mutex> guard(recvCntMutex_);
+    std::copy(msgRecvCnts_.begin(), msgRecvCnts_.end(), cnts.begin());
+  }
+
+  {
+    // copy sent message cnt to worldSize - 2*worldSize elements
+    std::lock_guard<std::mutex> guard(sendCntMutex_);
+    std::copy(
+        msgSendCnts_.begin(), msgSendCnts_.end(), cnts.begin() + worldSize);
+  }
```

Comment:
This can be turned into a simple class `MessageCounter` that deals with addition, the lock, and getting a snapshot (this is all identical between sending and receiving anyway). Then have 2 instances of this thing with an `append` function to add it to the same vector.

Instead of keeping the vector, this could also already be a tensor of size `[2, worldSize]`, to avoid the offsetting math below in the double for loop.